promql/engine.go

```diff
@@ -1540,6 +1541,18 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, annotations.Annotatio
 				}
 			}
 			ev.samplesStats.UpdatePeak(ev.currentSamples)
+
+			if e.Func.Name == "rate" || e.Func.Name == "increase" {
+				samples := inMatrix[0]
+				metricName := samples.Metric.Get(labels.MetricName)
+				if metricName != "" && len(samples.Floats) > 0 &&
```

Comment:
I forgot, maybe we talked about this, but I just noticed in usage that I only get the warning for an expression that actually has data - that is because the warning is added during evaluation time and not as a result of pure AST analysis after parsing the expression. Wouldn't the latter make more sense? I do something similar in PromLens: https://demo.promlens.com/?l=7PtzknhKgac (see the warning triangle next to the `rate()` in the tree UI).

Reply:
AST analysis would also be more efficient, I guess. On the other hand, it doesn't always work, e.g. `rate({__name__=~"foo.*"}[10m])` has to be evaluated first to find out if the annotation should happen or not.

I would say the current approach works in the general case. (If there is no data, we are also not doing anything wrong.) But ideally we might want to do AST analysis for static names and the current approach for the regexp case…

Reply:
That thought also crossed my mind, but note https://github.com/prometheus/prometheus/issues/13586 , i.e. we plan to make it a requirement to not return non-matching results.

Reply:
Those things are still possible. You just have to add labels to the results that fulfill the matchers sent. I don't think that's a big problem.

Reply:
Yeah, it's not super nice to return the whole SQL query again, but at least you can then choose to explicitly aggregate that label away before doing anything else with the result.